File: code/erlang/bbhw.erl
-module(bbhw).
-export([main/0, main/1]).

badArg(S) -> io:format("Invalid countdown ~p, try again...\n", [S]), main().
badNum(N) -> badArg(lists:flatten(io_lib:format("~p", [N]))).

main() ->
  case io:get_line("countdown: ") of
    eof        -> main();
    {error, _} -> main();
    Line       -> main(string:chomp(Line))
  end.

main([]) -> main();

main(F) when is_float(F)          -> badNum(F);
main(I) when is_integer(I), I < 0 -> badNum(I);
main(A) when is_atom(A)           -> main(atom_to_list(A));

main(L) when is_list(L) ->
  case io_lib:printable_list(L) of
    true ->
      case string:to_integer(L) of
        {I, []} -> main(I);
        _       -> badArg(L)
      end;
    false -> main(lists:concat(lists:join(" ", L)))
  end;

main(Count) when is_integer(Count), Count >= 0 -> 
  Loop = fun Loop(0) -> ok; Loop(I) ->
    io:format("~B...", [I]),
    timer:sleep(1000), 
    Loop(I-1)
  end,
  io:format("World, Hello..."),
  Loop(Count),
  io:format("Bye Bye.\n");

main(Unk) -> badArg(Unk).

File: code/elixir/bbhw.ex
defmodule BBHW do

  defp badArg(v), do: (IO.puts("Invalid countdown \"#{v}\", try again..."); main())
  
  def main() do
    case IO.gets("countdown: ") |> String.trim() do
      ""   -> main()
      line -> main(line)
    end
  end

  def main(""), do: main()

  def main(a) when is_atom(a),              do: badArg(a)
  def main(f) when is_float(f),             do: badArg(f)
  def main(i) when is_integer(i) and i < 0, do: badArg(i)
  def main(l) when is_list(l),              do: main(Enum.join(l, " "))

  def main(s) when is_binary(s) do
    case Integer.parse(s) do
      {i, ""} -> main(i);
      _       -> badArg(s)
    end
  end

  def main(count) when is_integer(count) and count >= 0 do
    loop = fn (0, _) -> :ok; (i, f) ->
      IO.write("#{i}...")
      Process.sleep(1000) 
      f.(i-1, f)
    end
    IO.write("World, Hello...")
    loop.(count, loop)
    IO.puts("Bye Bye.")
  end

  def main(Unk), do: badArg(Unk)
end

